### DISCUSSION TITLE
May I do the same thing with `typia` instead of using `zod`?

### DISCUSSION THREAD

**Discussion by samchon** [instructor-ai/instructor-js#89] [General]
I'm also utilizing the LLM with JSON schema like you in the same reason.

By the way, I am using [`typia`](https://github.com/samchon/typia) library instead of [`zod`](https://zod.dev). 

I understand why you've selected the [`zod`](https://zod.dev) library. It has huge ecosystem and loved by many TypeScript developers. 

However, I prefer to using the pure TypeScript type without any extra schema definition. Also, when many schema types be defined, [`zod`](https://zod.dev)  makes IDE extremely slow with enormous resource consumption, so that taking benefit of auto-completion by TypeScript types becomes impossible. Furthermore, I write description comments in as much detail as possible and use them actively. This is because this description is as important to the quality of LLM as the type or attribute name.

  - To use pure TypeScript type
  - To avoid IDE performance problem
  - To actively utilize description comments

In such reason, may I publish similar repo with you, and also link to your youtube video?

https://github.com/instructor-ai/instructor-js/assets/13158709/c00d3f8e-f709-41a9-9b42-d007f1e60928

```typescript
import typia, { tags } from "typia";

typia.json.application<[IBbsArticle]>();

/**
 * Article entity.
 *
 * `IBbsArticle* is a super-type entity of all kinds of articles in the current
 * backend system, literally shaping individual articles of the bulletin board.
 *
 * And, as you can see, the elements that must inevitably exist in the article,
 * such as the `title` or the `body`, do not exist in the `IBbsArticle`, but exist
 * in the subsidiary entity, {@link IBbsArticle.ISnapshot}, as a 1: N relationship,
 * which is because a new snapshot record is published every time the article is
 * modified.
 *
 * The reason why a new snapshot record is published every time the article is
 * modified is to preserve the evidence. Due to the nature of e-community, there
 * is always a threat of dispute among the participants. And it can happen that
 * disputes arise through articles or {@link IBbsArticleComment comments}, and to
 * prevent such things as modifying existing articles to manipulate the situation,
 * the article is designed in this structure.
 *
 * In other words, to keep evidence, and prevent fraud.
 *
 * @author Samchon
 */
interface IBbsArticle {
  /**
   * Primary Key.
   */
  id: string & tags.Format<"uuid">;

  /**
   * Writer of article.
   */
  writer: string;

  /**
   * List of snapshot contents.
   *
   * It is created for the first time when an article is created, and is
   * accumulated every time the article is modified.
   */
  snapshots: IBbsArticle.ISnapshot[] & tags.MinItems<1>;

  /**
   * Creation time of article.
   */
  created_at: string & tags.Format<"date-time">;
}
```

![image](https://github.com/instructor-ai/instructor-js/assets/13158709/9504aa48-4a46-4d46-9526-ee3ed5c18b75)

**Comment by jxnl** [comment]
I really recommend checking out TypeChat. https://github.com/microsoft/TypeChat

**Comment by r4kh4t** [comment]
@jxnl i wonder whether other schema libraries can be supported? I am using Joi in my backend, and will be more convenient if instructor can support this

**Reply by damusix** [reply]
+1, I also use Joi heavily and would be interested in something of this sort.

**Comment by samchon** [comment]
https://typia.io/docs/llm/application/

https://nestia.io/docs/swagger/chat/

You can directly make LLM function calling schema by `typia.llm.application()` function.

Also, it would better to utilizing the Swagger (OpenAPI) document as massive function providers.
